### PR TITLE
Use generic interfaces for Head, Chain Id, and Sequence

### PR DESCRIPTION
**Pull request by prashantkumar1982**
Txmgr was still importing native types for Head, chain-id and nonces.
Changing to generics now.

https://smartcontract-it.atlassian.net/browse/BCI-1159

### CODE REVIEW

File: common/txmgr/types/chain.go
Comment (jmank88) [inline common/txmgr/types/chain.go:7]: Can we keep our exceptional naming limited to type parameters?
```suggestion
// Sequence represents constraint on a chain's sequence object.
// It should be convertible to a string
type Sequence fmt.Stringer
```

Comment (jmank88) [reply common/txmgr/types/chain.go:7]: These actually aren't constraints as defined either. We use them to constrain type parameters, but they are just interface types that can be used like any other.

File: core/chains/evm/types/nonce.go
Comment (jmank88) [inline core/chains/evm/types/nonce.go:10]: Is this possible? It would be more efficient:
```suggestion
type Nonce int64
```

File: core/chains/evm/types/nonce.go
Comment (jmank88) [inline core/chains/evm/types/nonce.go:13]: nit/ unless this is on a hot path:
```suggestion
	return strconv.FormatInt(n.N, 10)
```

File: common/txmgr/types/mocks/utils.go
Comment (jmank88) [inline common/txmgr/types/mocks/utils.go:14]: Do we have to list all of these types out individually? Or could we infer some of them? e.g. given a tx, doesn't that tell us tx_hash and sequence type implicitly?

Comment (jmank88) [reply common/txmgr/types/mocks/utils.go:14]: I think they would still all have to be listed out on the `type TxStore[...]`, but it should be possible to omit a handful of them at call sites, however we'd have to constrain the types more strictly - not sure if that is helpful or not.

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go:10]: I tried it initially. But doesn't work.
Any caller that expects an int64 nonce, cannot use Nonce.

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go:13]: Done.

File: common/txmgr/types/mocks/utils.go
Comment (prashantkumar1982) [reply common/txmgr/types/mocks/utils.go:14]: I didn't quite understand what changes you are suggesting.
If we have TxStore type using x number of generic types, then each call site has to specify the native types there, no?
How can we omit certain types?

Also, I didn't get what you mean by "infer some of them".

File: common/txmgr/types/chain.go
Comment (prashantkumar1982) [reply common/txmgr/types/chain.go:7]: Done

File: common/txmgr/types/mocks/utils.go
Comment (jmank88) [reply common/txmgr/types/mocks/utils.go:14]: They don't always need to be explicitly listed. You can often omit the whole set: https://go.dev/blog/intro-generics#type-inference

Comment (jmank88) [reply common/txmgr/types/mocks/utils.go:14]: This doesn't seem like a good case for it, but this collection of types are connected enough that there could be other opportunities.

Comment (prashantkumar1982) [reply common/txmgr/types/mocks/utils.go:14]: Ohk, get it now.
Yes, it can apply for cases which match this criterion << Function argument type inference only works for type parameters that are used in the function parameters.

We could use it when we have function params matching the function argument types.

File: common/txmgr/types/chain.go
Comment (jmank88) [inline common/txmgr/types/chain.go:11]: ID should still be caps according to Go naming, since it's an acronym: 
```suggestion
// ID represents the base type, for any chain's ID.
// It should be convertible to a string, that can uniquely identify this chain
type ID fmt.Stringer
```

File: core/chains/evm/txmgr/eth_resender.go
Comment (jmank88) [inline core/chains/evm/txmgr/eth_resender.go:35]: Maybe out of scope for this PR, but have we considered favoring the 'local' `package types` when naming these? e.g.:
```suggestion
type EthResender[CHAIN_ID types.Id, ADDR common.Hashable[ADDR], TX_HASH common.Hashable[TX_HASH], BLOCK_HASH common.Hashable[BLOCK_HASH], SEQ types.Sequence] struct {
```
Excess packages and generic package names are usually anti-patterns in Go because of these sorts of naming collisions.

File: core/chains/evm/txmgr/reaper_test.go
Comment (jmank88) [inline core/chains/evm/txmgr/reaper_test.go:77]: find/replace typo? and above too?
```suggestion
		// Didn't delete because eth_tx has chain ID of 0
```

File: core/chains/evm/types/nonce.go
Comment (jmank88) [reply core/chains/evm/types/nonce.go:13]: Can you please elaborate? If you define `type Nonce int64`, then any `var n Nonce` can be trivially converted directly `int64(n)`.

File: core/services/vrf/integration_v2_test.go
Comment (jmank88) [inline core/services/vrf/integration_v2_test.go:508]: Lots of find/replace in this file:
```suggestion
// and asserts that the request ID logged by the RandomWordsRequested event
// matches the request ID that is returned and set by the consumer contract.
// The request ID and request block number are then returned to the caller.
```

File: core/chains/evm/types/models.go
Comment (jmank88) [inline core/chains/evm/types/models.go:467]:
```suggestion
			return errors.New("function ID must be 4 bytes in length")
```

File: core/chains/evm/txmgr/eth_resender.go
Comment (prashantkumar1982) [reply core/chains/evm/txmgr/eth_resender.go:35]: The chains/evm/txmgr and common/txmgr packages will merge soon, when we finish this refactoring, so this issue will be moot then.

File: common/txmgr/types/chain.go
Comment (prashantkumar1982) [reply common/txmgr/types/chain.go]: done

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go:13]: image

File: core/chains/evm/txmgr/reaper_test.go
Comment (prashantkumar1982) [reply core/chains/evm/txmgr/reaper_test.go]: Ohh, sorry about those. I used Goland's refactor option, didn't realize it will change all comments everywhere too.
Used the same refactor option again, to change Id to ID.

File: core/chains/evm/types/nonce.go
Comment (jmank88) [reply core/chains/evm/types/nonce.go:13]: You can convert it directly, in place or via a helper method:
image

File: common/txmgr/types/mocks/utils.go
Comment (aalu1418) [inline common/txmgr/types/mocks/utils.go:1]: is it possible to move these mock definitions to the evm/txmgr package?

It feels like we're getting close to a circular dependency by having the txmgr package import the common/txmgr and then have the common/txmgr/mocks import the evm/txmgr

Comment (aalu1418) [reply common/txmgr/types/mocks/utils.go:1]: at the very least I don't think this belongs in the common module - it will make it harder to move in the future since it would explicitly depend on a chain-specific component 😬

File: core/chains/evm/txmgr/attempts.go
Comment (aalu1418) [inline core/chains/evm/txmgr/attempts.go:109]: +1 to Jordan's comment about using a helper - it's cleaner than pulling the internal type and then wrapping it to ensure the types line up

File: core/chains/evm/txmgr/eth_confirmer.go
Comment (aalu1418) [inline core/chains/evm/txmgr/eth_confirmer.go:1119]: just curious, why are we changing logic? i thought this PR was only for type changes

File: core/services/keeper/upkeep_executer_test.go
Comment (aalu1418) [inline core/services/keeper/upkeep_executer_test.go:38]: this can use the mock type defined in txmmocks

File: core/web/resolver/resolver_test.go
Comment (aalu1418) [inline core/web/resolver/resolver_test.go:59]:
```suggestion
	txmStore    *txmgrtypesMocks.MockEvmTxStore
```

File: core/web/resolver/resolver_test.go
Comment (aalu1418) [inline core/web/resolver/resolver_test.go:35]: this can be removed

Comment (prashantkumar1982) [reply core/web/resolver/resolver_test.go:35]: Done

File: core/chains/evm/txmgr/eth_confirmer.go
Comment (prashantkumar1982) [reply core/chains/evm/txmgr/eth_confirmer.go:1119]: This for loop iterates and increments nonces. NextNonce = previousNonce + 1
This assumes nonces are integers, and next one is +1.
Once you generalize nonce into a generic Sequence type, we can no longer trivially get the next Sequence from a previous Sequence.
Previously this function received startingNonce and endingNonce values, and created all nonces. But now it can't.
So the caller has to pass in all Sequence objects in an array.

Another way could be, for the Sequence type to expose another function, saying getNextSequence(). But that's too complex, and may require even more inputs. For example, given a Bitcoin UTXO, you cannot trivially compute the next UTXO without first consuming current one with a Tx.


So overall, this above workaround side-steps this problem.

File: common/txmgr/types/mocks/utils.go
Comment (prashantkumar1982) [reply common/txmgr/types/mocks/utils.go:1]: Yes, I investigated many options before settling here.
Some points:
1) evm/txmgr and common/txmgr will merge into 1 package when this work ends
2) mock package will ALWAYS depend on parent type. So txmgr/mock will always depend on txmgr package, so any utility that you create for this Mock, has to depend on txmgr too, and thus that utility can't reside inside txmgr package.
3) The mock utils we create here, naturally depends on common/txmgr, but since we are instantiating with evm types, it also depends on evm packages.

Finally these are dependency constraints we will always have in future too(after evm/txmgr merges with common/txmgr):
common/txmgr/mock -->  common/txmgr
evm/...  --> common/txmgr

And our evm specific mock util will depend on all 3 packages above, i.e. common/txmgr/mock, common/txmgr, evm/...
So it has to be either in a separate package of its own, or can reside inside evm/... or txmgr/mock.

At present, I can't put it inside evm/... component, because cyclic deps.

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go:13]: Tagging @aalu1418 too for another comment on this same thing.

But is that option better than current?

Also, the helper methods don't work generically.


The below code still fails:
```

type Nonce int64

func (n Nonce) Int64() int64 {return int64(n)}

func DoSomething(n *int64) {
	// blah
}

func main() {
	var n Nonce = 12
	strconv.FormatInt(n.Int64(), 10)

	DoSomething(&n.Int64())
}


```

Comment (aalu1418) [reply core/chains/evm/types/nonce.go:13]: afaik you usually cannot create a new "object" and then take the reference of it immediately - that's just golang

also similar question to the address + hash types, why are we passing the nonce as a reference?

File: common/txmgr/types/mocks/utils.go
Comment (aalu1418) [reply common/txmgr/types/mocks/utils.go:1]: not sure I fully agree with all your points...the chain-agnostic logic from evm/txmgr will move to common/txmgr, and then evm/txmgr will import common/txmgr and provide the chain-specific types

regardless I don't think this is right place to put it (even if it's temporary), we are introducing evm specific types into a common module - could we put it in testutils or something like that?

File: core/chains/evm/txmgr/eth_confirmer.go
Comment (aalu1418) [reply core/chains/evm/txmgr/eth_confirmer.go:1119]: makes sense to accommodate generic types 👍🏼

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go:13]: I know you cannot take ref of a temporary object. So we have to do this:
```
nVal := n.Int64()
DoSomething(&nVal)
```

Instead, in the other solution, you just need single line:
`DoSomething(&n.N)`

Isn't the second option more concise?


In any way, I don't have any strong opinions here, and feel this is all just nit. I don't understand why it is more efficient to use "type Nonce int64".

File: core/chains/evm/txmgr/eth_confirmer.go
Comment (jmank88) [reply core/chains/evm/txmgr/eth_confirmer.go:1119]: Eventually, we may be able to use range over custom iterator types: https://github.com/golang/go/discussions/56413

File: core/chains/evm/types/nonce.go
Comment (jmank88) [reply core/chains/evm/types/nonce.go:13]: It is more concise, but it is meaningfully different to pass a pointer to internals of another type - do we actually need to do that for some kind of mutation case? If not, using a pointer doesn't generally make sense for primitive types that can be passed by value, using the same or less memory than the pointer itself.

File: common/txmgr/types/mocks/utils.go
Comment (prashantkumar1982) [reply common/txmgr/types/mocks/utils.go:1]: Ok, I think where we differ is that once chain agnostic code goes away from evm/txmgr, I assume there won't be any need for evm/txmgr package. It will all just go into evm/ package.

I now moved these utils to evm/txmgr/mocks package. This may still be a temporary place.

File: core/chains/evm/types/nonce.go
Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go]: Yes, we do need a ref to nonce here:
https://github.com/smartcontractkit/chainlink/blob/4c63c2a3f1cff445ab0d692e4fefa9f6225fdeb8/core/chains/evm/txmgr/models.go#L186

In any case, I don't have strong opinions, so switched to what you're suggesting. But at above place, it adds more conversion steps, and also, have to be careful while using it in SQL queries.
The DB SQL calls accept interface{}, but if the type doesn't match what they expect, they fail in runtime. So easy to make mistake passing Nonce where int64 was expected.

File: core/chains/evm/txmgr/eth_broadcaster_test.go
Comment (jmank88) [inline core/chains/evm/txmgr/eth_broadcaster_test.go:1929]: nit/ can this conversion be omitted? Plain `42` is an [untyped constant](https://go.dev/ref/spec#Constants) that should be directly convertible to `Nonce` IIUC
```suggestion
	require.Error(t, ethKeyStore.IncrementNextSequence(evmtypes.NewAddress(keyState.Address.Address()), &cltest.FixtureChainID, evmtypes.Nonce(42)))
```

File: core/chains/evm/types/nonce.go
Comment (jmank88) [inline core/chains/evm/types/nonce.go:9]: Does this compile with the pointer type? Given the value type receivers `(n Nonce)`, I would expect this:
```suggestion
var _ fmt.Stringer = zeroNonce
```

Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go]: Hmm, yes it was compiling with pointer type.
But now changed to 
var _ fmt.Stringer = Nonce(0)

Thanks!

**Review by jmank88 — DISMISSED** [review]

File: core/chains/evm/types/nonce.go
Comment (jmank88) [reply core/chains/evm/types/nonce.go]: Given:
```go
type Nonce int64
```
Since you can already do this directly:
```go
var n Nonce
_ = int64(n)
```
You could define the method to return a pointer:
```go
func (n Nonce) Int64() *int {
	return (*int)(&n)
}
```
No extra lines, and this has the benefit of returning a pointer to a newly allocated variable, rather than the one inside the original Nonce.

Comment (prashantkumar1982) [reply core/chains/evm/types/nonce.go]: Ok, nice, yes, that works.
So there are many places which require int64 and not *int64. So if I change the above to *int64, I will have to change all other places too.

Also thinking more about above Nonce *int64 type, that struct(EthTx) is also getting generalized in a future PL, into a SEQ type. So then we no longer will need a reference to Nonce.

Thus keeping it as it is for now, is good.

File: core/chains/evm/txmgr/eth_broadcaster_test.go
Comment (jmank88) [inline core/chains/evm/txmgr/eth_broadcaster_test.go:1064]: Can this be converted in one step?
```suggestion
	return uint64(n)
```

File: core/chains/evm/txmgr/evm_tx_store.go
Comment (jmank88) [inline core/chains/evm/txmgr/evm_tx_store.go:1364]: Are these necessary?
```suggestion
		res, err := tx.Exec(`UPDATE evm_key_states SET next_nonce = $1, updated_at = $2 WHERE address = $3 AND next_nonce = $4 AND evm_chain_id = $5`, newNextNonce, time.Now(), address, currentNextNonce, chainID.String())
```

File: core/chains/evm/txmgr/eth_broadcaster_test.go
Comment (prashantkumar1982) [reply core/chains/evm/txmgr/eth_broadcaster_test.go:1064]: thanks!

File: core/chains/evm/txmgr/evm_tx_store.go
Comment (prashantkumar1982) [reply core/chains/evm/txmgr/evm_tx_store.go:1364]: Ya I wanted to be safe here, since the tx.Exec() accepts interface{} for every argument. I don't know if the validation will accept Nonce instead of int64. The failure, if detected, will only be at runtime, and hard to catch in cases of tests which don't bubble up errors correctly.

**Review by jmank88 — APPROVED** [review]

**Pull request merged**
